### PR TITLE
Replumbing: pre-populate the EVP namemap with commonly known names

### PR DESCRIPTION
**Pull request by levitte**
This adds ossl_namemap_empty(), to detect if a namemap is empty and
can thereby be pre-populated.

Related to #8967

### CODE REVIEW

File: crypto/evp/ossl_kdf_names.h [hunk 1]
```diff
@@ -0,0 +1,52 @@
+/*
+ * Copyright 2019 The OpenSSL Project Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License 2.0 (the "License").  You may not use
+ * this file except in compliance with the License.  You can obtain a copy
+ * in the file LICENSE in the source distribution or at
+ * https://www.openssl.org/source/license.html
+ */
+
+/* ONLY included by evp_fetch.c */
```
Comment (richsalz) [inline crypto/evp/ossl_kdf_names.h]: Maybe .h is not the best name for what's really C code, especially if it should only be included ONCE?  .inc perhaps?

File: crypto/evp/ossl_kdf_names.h [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+
+#include <openssl/obj_mac.h>
+
+/* Stuff from c_allkdf.c */
+static const char * const id_pbkdf2_names[] =
+    { LN_id_pbkdf2, NULL };
+
+
```
Comment (richsalz) [inline crypto/evp/ossl_kdf_names.h]: any reason for double blank lines?

File: crypto/core_namemap.c [hunk 3]
```diff
@@ -11,44 +11,133 @@
...
+ * | "namen"        | <------+
+ * +----------------+
+ *
+ * These are then used in a stack indexed by number (for number to
+ * names mapping)
+ */
+
 typedef struct {
+    /* names and n_names are organized the same way as argv and argc */
+    const char **names;
```
Comment (paulidale) [inline crypto/core_namemap.c]: Would make more sense to use a stack here?

Perhaps also for the body but that's slightly further from the stack behaviour.

**Comment by levitte** [comment]
@paulidale, it would probably be more constructive to discuss the technical aspects of this PR here...

**Comment by levitte** [comment]
This includes #9969, please review that one first

**Comment by levitte** [comment]
I'm putting this on hold, because it should be reworked to work together with the changes in #8985, i.e. that one needs to be merged first

**Comment by levitte** [comment]
(the hold label that I added was on a personal basis, not an OMC decision thing, so I removed it)

**Comment by levitte** [comment]
So, this is essentially finished at this point.  Updating it was a breeze...

However, I wonder if this is still desirable.  We have lived without a pre-populated namemap for a while now, and I still see no signs of panic or horror with regards to that, so ...

**Comment by mattcaswell** [comment]
I think this is a valuable thing to have. We want to make it as easy as possible for people to transition from one provider to another. At the moment all of our aliases are held in the default provider - but if you don't load that then you don't get the full list of aliases. Having them pre-populated in the namemap would mean we always have them.

**Comment by levitte** [comment]
I put this in WIP because @mattcaswell and I realised that we need to keep the legacy method database (which is often wrongly called "object database") around, and I might as well use that to pre-populate the EVP namemap instead of having an intermediary array with the exact same information.

**Comment by levitte** [comment]
There's something very odd happening with the solution to pre-populate names from the legacy method db.

**Comment by levitte** [comment]
It seems I finally pulled through all corner cases (and there were a few).  If the CIs agree, I'll take this out of WIP.

**Comment by levitte** [comment]
Not WIP any more

**Comment by levitte** [comment]
Ping

File: doc/internal/man3/ossl_namemap_new.pod [hunk 4]
```diff
@@ -40,6 +41,10 @@ new B<OSSL_NAMEMAP>.
 This is suitable to use when the B<OSSL_NAMEMAP> is embedded in other
 structures, or should be independent for any reason.
 
+ossl_namemap_empty() checks if the given B<OSSL_NAMEMAP> is empty or
+not.
+This can be useful for pre-population purposes.
```
Comment (paulidale) [inline doc/internal/man3/ossl_namemap_new.pod]: I'd omit this line, it's describing a use rather than the function.

**Review by paulidale — APPROVED** [review]

File: doc/internal/man3/ossl_namemap_new.pod (hunk 4 above)
Comment (levitte) [reply doc/internal/man3/ossl_namemap_new.pod]: Okie

**Comment by levitte** [comment]
Merged.

6a835fcfb10ba004498f9e39873db3d2b9011609 Replumbing: pre-populate the EVP namemap with commonly known names

**Pull request closed without merging**